doc/articles/Concepts/Authentication.md

```diff
@@ -16,67 +16,189 @@ When you have a multiplayer game, often you need to store information about your
+
+    -   Remember to put a small delay on the Disconnect call on the server if you send them a failure message so that it has a chance to be delivered before the connection is dropped.
+
+-   `NetworkConnection` has an `AuthenticationData` object where you can drop a class instance of any data you need to persist on the server related to the authentication, such as account id's, tokens, character selection, etc.
+
+Now that you have the foundation of a custom Authenticator component, the rest is up to you. You can exchange any number of custom message between the server and client as necessary to complete your authentication process before approving the client.
+
+## Basic Authentication
+
+To get you started, here's a complete example of a custom Authenticator.
+
+-   Create a new script in your project (not in the Mirror folders) called `BasicAuthenticator`
+
+-   Replace the boilerplate code that Unity provides with the code below and save
+
+-   Drag the script to the inspector of the object in your scene that has Network Manager
+
+-   Assign the Basic Authenticator component to the Authenticator field in Network Manager
+
+-   Remove the Default Authenticator component.
+
+When you're done, it should look like this:
+
+![Inspector showing Basic Authentication component](BasicAuthentication.PNG)
 
-4.  Override the `OnServerAddPlayer` method and validate the user's credential.
+>   **Note:** You don't need to assign anything to the event lists unless you want to subscribe to the events in your own code for your own purposes. Mirror has internal listeners for both events.
 
-For example this would be part of your `NetworkManager` class:
+### Basic Authenticator
 
-```cs
-class MyGameNetworkManager : NetworkManager
+``` cs
+using UnityEngine;
+using Mirror;
+
+public class BasicAuthenticator : Authenticator
 {
-    class CredentialsMessage : MessageBase
+    [Header("Custom Properties")]
+
+    // for demo purposes, set these in the inspector
+    public string username;
+    public string password;
+
+    public class AuthRequestMessage : MessageBase
     {
         // use whatever credentials make sense for your game
         // for example, you might want to pass the accessToken if using oauth
-        public string username;
-        public string password;
+        public string authUser;
+        public string authPass;
+    }
+
+    public class AuthResponseMessage : MessageBase
+    {
+        public byte code;
+        public string message;
     }
 
-    // this gets called on the client after it has connected to the server
-    public override void OnClientConnect(NetworkConnection conn)
+    public void Start()
     {
-        base.OnClientConnect(conn);
+        NetworkServer.RegisterHandler<AuthRequestMessage>(OnAuthRequestMessage);
+        NetworkClient.RegisterHandler<AuthResponseMessage>(OnAuthResponseMessage);
```

Comment:
This won't work with #1067 
Once you disconnect,  if you try to connect again,  this will no longer be registered

Reply:
Closed 1067

Reply:
Actually I was thinking to move these to ClientAuthenticate and ServerAuthenticate

Reply:
ok, that's a doc change...let's not block the merge and I'll revisit the doc.